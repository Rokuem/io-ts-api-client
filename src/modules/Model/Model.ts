import * as t from 'io-ts';
import reporter from 'io-ts-reporters';
import { Either, isRight } from 'fp-ts/lib/Either';
import { TypedEmitter } from '../TypedEmitter/TypedEmmiter';
import { addedDiff } from 'deep-object-diff';
import { ValidationOptions } from '../types';
import { deepMerge } from '../../helpers/deepMerge';

/**
 * Simple io-ts model with a validator. This is used to safeguard against API changes and to provide a easy solution for type checking it.
 */
export class Model<Schema extends t.Any = any, SampleKey extends string = never, Samples extends t.TypeOf<Schema> = t.TypeOf<any>> {
  /**
   * Emitter for modal events.
   */
  public static emitter = new TypedEmitter<{
    'before-validation': (model: string) => void;
    'validation-success': (model: string) => void;
    'validation-error': (model: string, error: Error) => void;
    'after-validation': (model: string) => void;
    'extra-keys-detected': (model: string, error: Error) => void;
  }>();

  /**
   * a name to call the model during logs and errors.
   */
  public name: string;

  /**
   * @deprecated - use model.schema instead.
   */
  public base: Schema;

   /**
   * IO-TS interface used for the model.
   *
   * you can use this to extend other interfaces.
   */
  public schema!: Schema;

  /**
   * Ts interface of the model.
   * 
   * !Warn: Do not use this as an value.
   * @since 0.7.4
   */
  public tsInterface!: t.TypeOf<Schema>

  public samples!: Record<SampleKey, Samples>;

  public constructor(config: {
    /**
     * The name of the model.
     */
    name: string;
    /**
     * The io-ts model.
     */
    schema: Schema;
    samples?: Record<SampleKey, Samples>;
  }) {
    this.name = config.name;
    this.base = config.schema;
    this.schema = config.schema;
    this.samples = config.samples || {} as any;
  }

  /**
   * If not an empty string, this model belongs to an operation.
   */
  public operation = '';

  /**
   * Checks if there are any extra keys in the response received.
   */
  private checkForExtraKeys(options: {
    result: Either<t.Errors, any>;
    target: any;
    debug: boolean;
    throwErrors: boolean;
  }) {
    const { result, target } = options;

    if (!('right' in result) && options.debug) {
      console.warn('Extra fields check skipped due to failed validation');
      return;
    }

    if ('props' in this.base && 'right' in result) {
      const filteredDecode = t.exact(this.base as any).decode(target);
      const resultJson = JSON.stringify(result.right);

      if (!('right' in filteredDecode)) {
        throw new Error('Something went wrong while checking for extra keys');
      }

      const filtered = filteredDecode.right;
      const filteredJson = JSON.stringify(filtered);

      if (filteredJson !== resultJson) {
        const diff = addedDiff(filtered, result.right);
        const errorMessage = `${
          this.operation ? `[${this.operation}] ` : ''
        }Detected extra properties in model "${this.name}": ${JSON.stringify(
          diff,
          null,
          '  '
        )}`;

        Model.emitter.emit('extra-keys-detected', this.name, new Error(errorMessage));

        if (options.throwErrors) {
          throw new Error(errorMessage);
        } else if (options.debug) {
          console.error(errorMessage);
        }
      }

      return filtered;
    }
  }

  private handleInvalidResult(options: {
    result: Either<t.Errors, any>;
    target: any;
    throwErrors: boolean;
    debug: boolean;
  }) {
    const validationErrors = reporter.report(options.result);

    for (const error of validationErrors) {
      const validationMessage = `ERROR - ${
        this.operation ? `[${this.operation}]` : ''
      }[${this.name}] VALIDATION FAILED: ${error} at: \n ${JSON.stringify(
        options.target,
        null,
        2
      )}`;

      Model.emitter.emit('validation-error', this.name, new Error(validationMessage));

      if (options.debug) {
        console.error(validationMessage);
      }

      if (options.throwErrors) {
        throw new Error(validationMessage);
      }
    }
  }

  private handleValidResult(
    target: any,
    options: ValidationOptions
  ) {
    if (options.debug) {
      console.log(
        `SUCCESS - ${this.operation ? `[${this.operation}]` : ''}[${
          this.name
        }]: Target is valid!`,
        {
          target,
        }
      );
    }

    Model.emitter.emit('validation-success', this.name);
  }

  /**
   * Validates the model. Throws error if failed in strict mode.
   */
  public validate(
    target: any,
    options: ValidationOptions
  ) {
    Model.emitter.emit('before-validation', this.name);

    let result = this.base.decode(target);
    const isValid = isRight(result);

    if (options.strictTypes) {
      const filtered = this.checkForExtraKeys({
        result,
        target,
        debug: options.debug,
        throwErrors: options.throwErrors,
      });

      if (filtered) {
        result = {
          ...result,
          right: filtered,
        } as any;
      }
    }

    if (!isValid) {
      try {
        this.handleInvalidResult({
          target,
          result,
          debug: !!options.debug,
          throwErrors: !!options.throwErrors,
        });
      } catch (e) {
        Model.emitter.emit('after-validation', this.name);
        throw e;
      }
    } else {
      this.handleValidResult(target, options);
    }

    Model.emitter.emit('after-validation', this.name);

    return 'right' in result ? result.right : target;
  }

  /**
   * Asserts that target is valid for this model.
   */
  public assert(target: any, options?: { 
    /**
     * Fails the assertion in case the target has extra properties.
     */
    strict?: boolean;
    /**
     * Debug the validation process.
     */
    debug?: boolean;
  }): target is ModelInterface<this> {
    try {
      this.validate(target, {
        debug: !!options?.debug,
        throwErrors: true,
        strictTypes: !!options?.strict,
      });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Creates an object that matches the model.
   * Useful for creating mocks.
   */
  public createSample<T extends t.TypeOf<Schema>>(sample: T) {
    return sample;
  }

  /**
   * Extend one of the declared samples in the model declaration.
   * Useful for tests.
   */
  public extendSample<T extends Partial<t.TypeOf<Schema>>, K extends SampleKey>(base: K, sample: T) {
    return deepMerge(this.samples[base], sample);
  }
}

export type ModelInterface<T extends Model<any>> = t.TypeOf<T['base']>;
