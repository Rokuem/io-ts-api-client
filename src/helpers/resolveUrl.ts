export function addPathToUrl(url: URL, ...additions: string[]) {
  for (const addition of additions) {
    if (url.pathname.endsWith('/')) {
      if (addition.startsWith('/')) {
        url.pathname += '/' + addition;
      } else {
        url.pathname += addition;
      }
    } else {
      if (addition.startsWith('/')) {
        url.pathname += addition;
      } else {
        url.pathname += '/' + addition;
      }
    }
  }
}
