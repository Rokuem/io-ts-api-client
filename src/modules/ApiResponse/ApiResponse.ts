import { Model } from '../Model/Model';
import { HttpStatus } from '../../constants/HttpStatus';

/**
 * Describes a possible response for the resource.
 */
export class ApiResponse<M extends Model, S extends HttpStatus> {
  public model!: M;
  public status!: S;
  constructor(options: { model: M; status: S }) {
    Object.assign(this, options);
  }
}
