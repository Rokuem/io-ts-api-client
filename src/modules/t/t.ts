import * as types from 'io-ts';
import * as extraTypes from 'io-ts-types';

export const t = {
  ...types,
  ...extraTypes,
  /**
   * Creates null intersection with provided type.
   */
  nullable: <T extends types.Any>(type: T) => types.union([t.null, type]),
  /**
   * Creates intersection between required and optional props.
   * @since 0.7.4
   */
  schema: <T extends types.Props, P extends types.Props>(type: { required: T, optional: P }) => types.intersection([
    types.interface(type.required),
    types.partial(type.optional)
  ]),
  /**
   * Type for valid date strings.
   * @since 0.8.5
   */
  dateString: new types.Type<string, string, unknown>(
    'date string',
    (input: unknown): input is string => typeof input === 'string',
    (input, context) => {
      if (typeof input !== 'string') {
        return types.failure(input, context);
      }

      const date = new Date(input);

      if (isNaN(date.getTime())) {
        return types.failure(input, context);
      }

      return types.success(input);
    },
    types.identity
  ),
  /**
   * Type for valid urls.
   * @since 0.8.5
   */
  url: new types.Type<string, string, unknown>(
    'url',
    (input: unknown): input is string => typeof input === 'string',
    (input, context) => {
      if (typeof input !== 'string') {
        return types.failure(input, context);
      }
  
      try {
        // URL constructors fails with invalid urls.
        new URL(input);
        return types.success(input);
      } catch (e) {
        return types.failure(input, context);
      }
    },
    types.identity
  )
};