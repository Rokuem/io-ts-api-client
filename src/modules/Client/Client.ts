import axios from 'axios';
import { Resource } from '../Resource/Resource';
import { mapObject } from '../../helpers/mapObject';
import { addPathToUrl } from '../../helpers/resolveUrl';
import { ApiResponse } from '../ApiResponse/ApiResponse';
import { ResponseWithStatus } from '../../helpers/ResponseWithStatus.type';
import { ModelInterface } from '../Model/Model';

/**
 * API client constructor. Use this to define a client then, use the `client.getApi()` to get the typesafe api.
 *
 * After declaring the resources you will be able to execute its operations with the client API.
 *
 * @example
 *  const client = new Client({
 *    //...
 *  });
 *
 *  const API = client.getApi();
 *
 *  const res = await API.someResource.someOperation();
 *
 *  if (res.status === HTTPStatus.OK) { // typeguard
 *    console.log(res.data)
 *  }
 */
export class Client<
  Resources extends Record<string, Resource<any, any>>,
  GlobalResponses extends ApiResponse<any, any> = ApiResponse<never, never>
> {
  /**
   * Axios instance this client should use.
   *
   * This will be created by default, but can be modified.
   *
   * @default axios.create()
   */
  public axiosInstance = axios.create();
  /**
   * URL object based on the client base url.
   */
  public base: URL;

  /**
   * List of resources provided by the base url of the API.
   *
   * Each resource can have more paths associated to it.
   */
  private readonly resources: Resources;

  /**
   * Global responses that can be returned by the operations.
   */
  private globalResponses: [...GlobalResponses[]] = [];

  constructor({
    base,
    resources,
    globalResponses,
  }: {
    /**
     * URL of the server that will be used as a base.
     */
    base: string;
    /**
     * List of resources this url provides.
     */
    resources: Resources;
    /**
     * Responses that might be returned in any operation.
     */
    readonly globalResponses?: [...GlobalResponses[]];
  }) {
    try {
      this.base = new URL(base);
      this.resources = resources;
      this.globalResponses = globalResponses || [];
    } catch (e) {
      throw new Error(
        'Could not create client. Provided base is an invalid URL: ' + base
      );
    }
  }

  /**
   * Gets the typesafe client API for executing the operations.
   */
  public getApi() {
    type ResourceOperations<
      ResourceKey extends keyof Resources
    > = Resources[ResourceKey]['operations'];

    type Operation<
      ResourceKey extends keyof Resources,
      ResourceOperation extends keyof ResourceOperations<ResourceKey>
    > = Resources[ResourceKey]['operations'][ResourceOperation];

    type OperationOptions<
      ResourceKey extends keyof Resources,
      OperationKey extends keyof ResourceOperations<ResourceKey>
    > = Operation<ResourceKey, OperationKey>['options'] extends undefined
      ? []
      : [Operation<ResourceKey, OperationKey>['options']];

    type MapResponses<Responses extends [...ApiResponse<any, any>[]]> = {
      [key in keyof Responses]: Responses[key] extends ApiResponse<
        infer ResponseModel,
        infer ResponseStatus
      >
        ? Promise<
            ResponseWithStatus<ResponseStatus, ModelInterface<ResponseModel>>
          >
        : never;
    }[number];

    type API = {
      [ResourceKey in keyof Resources]: {
        [OperationKey in keyof ResourceOperations<ResourceKey>]: (
          ...args: OperationOptions<ResourceKey, OperationKey>
        ) => GlobalResponses extends ApiResponse<never, never>
          ? MapResponses<Operation<ResourceKey, OperationKey>['responses']>
          :
              | MapResponses<Operation<ResourceKey, OperationKey>['responses']>
              | MapResponses<[...GlobalResponses[]]>;
      };
    };
    return mapObject(this.resources, (resourceName, resource) => {
      return [
        resourceName,
        mapObject(resource.operations, (operationName) => [
          operationName,
          (options: any) => {
            const operationUrl = new URL(this.base.href);

            if (resource.basePath) {
              addPathToUrl(operationUrl, resource.basePath);
            }

            if (typeof operationName !== 'string') {
              throw new Error('Operation names must be string');
            }

            return resource.execute({
              operation: operationName,
              axios: this.axiosInstance,
              options,
              url: operationUrl,
              globalResponses: this.globalResponses,
            });
          },
        ]),
      ];
    }) as API;
  }
}
