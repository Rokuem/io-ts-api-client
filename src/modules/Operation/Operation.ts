import { AxiosError } from "axios";
import { HttpMethod } from "../../constants/httpMethod";
import { ApiResponse } from "../ApiResponse/ApiResponse";
import { Model } from "../Model/Model";
import { ModelInterface } from "../Model/Model";
import { ResponseWithStatus } from "../../helpers/ResponseWithStatus.type";
import { t } from "../t/t";
import { Client } from "../Client/Client";
import { Resource } from "../Resource/Resource";
import { ValidationOptions } from "../types";

/**
 * Operations describes a request.
 *
 * You can pass its options as the first type argument.
 */
export class Operation<
  Payload extends Model<any>,
  Responses extends [...ApiResponse<any, any>[]],
  Method extends HttpMethod,
  Options = never
> {
  /**
   * Name of the operation. Infered when the client is made.
   */
  public name = "Operation";
  /**
   * function to construct the url of the operation.
   *
   * @param url - The base URL object of the api.
   * @param options - The options of the operation.
   */
  public url!: (...args: [url: URL, options: Options]) => URL;
  /**
   * HTTP method of the operation.
   */
  public method!: Method;
  /**
   * IO-TS Model used to validate the payload.
   */
  public payloadModel?: Payload;
  /**
   * List of possible responses for the operation.
   *
   * This can be merged with global responses of a client.
   */
  public readonly responses!: [...Responses];
  /**
   * Options of the operation.
   *
   * This is just for typing so you can use `{} as Options` for example.
   * You can also provide this in the operations constructor.
   */
  public options?: Options;
  /**
   * Function to get the object of headers to send with the request.
   */
  public headers?:
    | Record<string, string>
    | ((options?: Options) => Record<string, string>);
  /**
   * Function to mock the response. Useful for testing and for development when the API is not ready.
   */
  public mock?: (
    /**
     * Computed url returned by the Operation url option.
     */
    url?: URL,
    /**
     * Payload returned by the PayloadConstructor
     */
    payload?: ModelInterface<Payload>
  ) => Promise<
    | Partial<
        {
          [key in keyof Responses]: Responses[key] extends ApiResponse<any, any>
            ? Partial<
                ResponseWithStatus<
                  Responses[key]["status"],
                  ModelInterface<Responses[key]["model"]>
                >
              >
            : Responses[key];
        }[number]
      >
    | false
  >;
  /**
   * Function to construct the payload using the provided options for the operation.
   */
  payloadConstructor?: (options?: Options) => ModelInterface<Payload>;
  constructor(
    config: Pick<
      Operation<Payload, Responses, Method, Options>,
      | "url"
      | "method"
      | "payloadModel"
      | "payloadConstructor"
      | "responses"
      | "options"
      | "headers"
      | "mock"
    >
  ) {
    Object.assign(this, config);
  }

  public setName(name: string) {
    this.name = name;

    for (const response of this.responses) {
      if (response.model instanceof Model) {
        response.model.operation = this.name;
      }
    }

    if (this.payloadModel) {
      this.payloadModel.operation = this.name;
    }
  }

  /**
   * Executes the operation using the provided configuration.
   */
  public async execute<GlobalResponses extends ApiResponse<any, any>>({
    axiosInstance,
    options,
    base,
    globalResponses,
    debug,
    throwErrors,
    strictTypes,
    globalHeaders,
  }: {
    globalHeaders: Record<string, string>;
  } & Pick<
    Client<
      Record<string, Resource<Record<string, this>, GlobalResponses>>,
      GlobalResponses
    >,
    "base" | "globalResponses" | "axiosInstance"
  > &
    ValidationOptions &
    Pick<this, "options">) {
    const log = (...message: any[]) => {
      if (debug) {
        console.log(...message);
      }
    };

    const logError = (...message: any[]) => {
      if (debug) {
        console.error(...message);
      }
    };

    if (this.payloadConstructor) {
      const payload = this.payloadConstructor?.(options);
      log("Payload: ", payload);
      this.payloadModel?.validate(payload, {
        debug,
        throwErrors,
        strictTypes,
      });
    }

    type GetResponses<T extends [...ApiResponse<any, any>[]]> = {
      [key in keyof T]: T[key] extends ApiResponse<any, any>
        ? ResponseWithStatus<T[key]["status"], ModelInterface<T[key]["model"]>>
        : T[key];
    }[number];

    type $Response =
      | GetResponses<Responses>
      | GetResponses<[...GlobalResponses[]]>;

    const requestUrl = this.url(base as any, options as any);

    if (this.mock) {
      log("mock detected!");
      const mock = await this.mock(
        requestUrl,
        this.payloadConstructor?.(options)
      );
      log("Mock: ", mock);
      if (mock !== false) {
        return { ...mock, body: (mock as any)?.data } as $Response;
      }
    }

    try {
      log("Making axios request");
      const response = await axiosInstance.request<$Response>({
        method: this.method as any,
        data: this.payloadConstructor?.(options),
        url: requestUrl.href,
        headers: {
          ...globalHeaders,
          ...(typeof this.headers === "function"
            ? this.headers?.(options)
            : this.headers || {}),
        },
        validateStatus: (status) =>
          this.responses.some((res) => res.status === status),
      });

      if (!response) {
        throw new Error(`[${this.name}] Received empty reponse`);
      }

      const matchingResponses = [...this.responses, ...globalResponses].filter(
        (res) => res.status === response.status
      );

      log("matching Responses: ", matchingResponses.length);

      let [responseDeclaration] = matchingResponses;

      if (matchingResponses.length > 1) {
        responseDeclaration = new ApiResponse({
          model: new Model({
            schema: t.intersection([
              ...matchingResponses.map((res) => res.model.base),
            ] as any),
            name: matchingResponses.map((res) => res.model.name).join(" | "),
          }),
          status: responseDeclaration.status,
        });
      }

      if (
        !responseDeclaration ||
        !(responseDeclaration.model instanceof Model)
      ) {
        console.error(`[${this.name}] UNEXPECTED RESPONSE: `, response);
        throw new Error(
          `Unexpected response without declaration. Status: ${
            response.status
          }, data: ${JSON.stringify(response.data)}`
        );
      }

      responseDeclaration.model.operation = this.name;

      const data = responseDeclaration.model.validate((response as any).data, {
        debug,
        strictTypes,
        throwErrors,
      });

      return {
        ...response,
        data,
        body: data,
      };
    } catch (error) {
      if (typeof error === "object" && error && "response" in error) {
        const axiosError = error as AxiosError;
        logError("Unexpected response: ", axiosError.response);
        axiosError.message = `MODEL NOT FOUND - Unexpected response received from operation ${
          this.name
        }: ${axiosError.response?.status} - ${JSON.stringify(
          axiosError.response?.data
        )}`;
      }

      throw error;
    }
  }
}
