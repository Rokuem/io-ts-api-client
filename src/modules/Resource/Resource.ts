import { Operation } from '../Operation/Operation';
import { addPathToUrl } from '../../helpers/resolveUrl';
import { ApiResponse } from '../ApiResponse/ApiResponse';
import type { Client } from '../Client/Client';
import { ValidationOptions } from '../types';

/**
 * A resource is a base of the API. It contains operations that will execute using it as a base.
 */
export class Resource<
  Operations extends Record<
    string,
    Operation<any, any, any, never> | Operation<any, any, any, any>
  >,
  GlobalResponses extends ApiResponse<any, any>
> {
  /**
   * Path to append to the client url for all operations.
   */
  public basePath = '';

  /**
   * Api operations the resource can perform.
   */
  public operations!: Operations;

  constructor(
    config: Partial<Pick<Resource<any, any>, 'basePath'>> & {
      operations: Operations;
    }
  ) {
    Object.assign(this, config);
    this.assignOperationNames();
  }

  private assignOperationNames() {
    for (const key in this.operations) {
      const operation = this.operations[key];
      operation.setName(key);
    }
  }

  /**
   * Executes an operation inside a resource.
   */
  public execute<K extends keyof Operations>(
    options: Pick<
      Client<
        Record<string, Resource<Operations, GlobalResponses>>,
        GlobalResponses
      >,
      'base' | 'globalResponses' | 'axiosInstance'
    > &
      Pick<Operations[K], 'options'> & {
        operation: K;
      } & ValidationOptions
  ) {
    const resourceUrl = new URL(options.base.href);

    if (this.basePath) {
      addPathToUrl(resourceUrl, this.basePath);
    }

    const operation = this.operations[options.operation];

    if (!(operation instanceof Operation)) {
      throw new Error('Invalid operation: ' + options.operation);
    }

    return operation.execute({ ...options, base: resourceUrl })
  }
}
