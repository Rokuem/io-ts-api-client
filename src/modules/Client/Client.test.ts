import mockAxios from 'jest-mock-axios';
import { expectTypeOf } from 'expect-type';
import { Client } from './Client';
import { testConfig } from '../../testConfig';
import { Resource } from '../Resource/Resource';
import { Operation } from '../Operation/Operation';
import { HttpMethod } from '../../constants/httpMethod';
import { addPathToUrl } from '../../helpers/resolveUrl';
import { ApiResponse } from '../ApiResponse/ApiResponse';
import { HttpStatus } from '../../constants/HttpStatus';
import { t } from '../t/t';
import Model from '../Model/Model';

const okSampleResponse = new ApiResponse({
  status: HttpStatus.OK,
  model: new Model({
    name: 'API sample resource',
    schema: t.interface({
      ok: t.boolean,
    }),
  }),
});

describe('A Client', () => {
  afterEach(() => {
    mockAxios.reset();
    Model.strictTypes = false;
    Model.throwErrors = false;
  });

  beforeAll(() => {
    Model.strictTypes = true;
    Model.throwErrors = true;
  });

  const client = new Client({
    base: testConfig.testServerUrl,
    globalResponses: [
      new ApiResponse({
        status: HttpStatus.INTERNAL_SERVER_ERROR,
        model: new Model({
          name: 'API 500 resource',
          schema: t.interface({
            ok: t.literal(false),
          }),
        }),
      }),
    ],
    resources: {
      samples: new Resource({
        operations: {
          getOk: new Operation({
            method: HttpMethod.GET,
            url(url) {
              addPathToUrl(url, '/sample/ok');
              return url;
            },
            responses: [okSampleResponse],
          }),
          getSample: new Operation({
            method: HttpMethod.GET,
            options: {} as {
              sampleType: 'ok' | 'accepted';
            },
            url: (url, { sampleType }) => {
              addPathToUrl(url, '/sample/' + sampleType);
              return url;
            },
            responses: [
              okSampleResponse,
              new ApiResponse({
                status: HttpStatus.ACCEPTED,
                model: new Model({
                  name: 'API sample resource',
                  schema: t.interface({
                    accepted: t.boolean,
                  }),
                }),
              }),
            ],
          }),
        },
      }),
      extraResource: new Resource({
        operations: {
          getOk: new Operation({
            method: HttpMethod.GET,
            url(url) {
              addPathToUrl(url, '/sample/ok');
              return url;
            },
            responses: [okSampleResponse],
          }),
        },
      }),
    },
  });

  test('Should match the snapshot', () => {
    expect(client).toMatchSnapshot('sample client');
  });

  describe('Its API', () => {
    const API = client.getApi();

    test('Should convert its operations to an api', async () => {
      const resource = 'samples';
      expect(API).toHaveProperty(resource);
      expectTypeOf(API).toHaveProperty(resource);

      const operation = 'getSample';
      expect(API.samples).toHaveProperty(operation);
      expectTypeOf(API.samples).toHaveProperty(operation);

      const expected = {
        sampleType: '' as 'ok' | 'accepted',
      } as const;

      expectTypeOf(API.samples.getSample).parameters.toMatchTypeOf([
        expected as typeof expected | undefined,
      ]);
      expectTypeOf(API.samples.getOk).parameters.toMatchTypeOf([] as const);
    });

    test('Should return results correctly', async () => {
      const request = API.samples.getOk();
      mockAxios.mockResponse({
        data: {
          ok: true,
        },
      });

      const res = await request;

      expect(res.data).toEqual({
        ok: true,
      });
      expectTypeOf(res.data).toMatchTypeOf({
        ok: true,
      });

      expect(res.status).toBe(HttpStatus.OK);
      expectTypeOf(res.status).toMatchTypeOf(HttpStatus.OK);

      const request2 = API.samples.getSample({
        sampleType: 'ok',
      });
      mockAxios.mockResponse({
        data: {
          sampleType: 'ok',
        },
      });

      const res2 = await request2;

      if (res2.status === HttpStatus.OK) {
        expectTypeOf(res2.data).toMatchTypeOf({
          ok: true,
        });
      }

      if (res2.status === HttpStatus.ACCEPTED) {
        expectTypeOf(res2.data).toMatchTypeOf({
          accepted: true,
        });
      }

      if (res2.status === HttpStatus.INTERNAL_SERVER_ERROR) {
        expectTypeOf(res2.data).toMatchTypeOf({
          ok: false,
        } as const);
      }
    });

    test('Should accept global responses', async () => {
      Model.strictTypes = true;
      Model.throwErrors = true;

      const request = API.samples.getOk();
      const mockedRes = {
        status: HttpStatus.INTERNAL_SERVER_ERROR,
        data: {
          ok: false,
        },
      };

      mockAxios.mockResponse(mockedRes);

      expect(await request).toEqual(expect.objectContaining(mockedRes));
    });
  });
});
