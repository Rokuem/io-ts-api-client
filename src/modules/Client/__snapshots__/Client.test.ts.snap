// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Client Its API When executing an operation When the response status is not declared The operation should throw and error: Missing Model Error 1`] = `[Error: Unexpected response without declaration. Status: 333, data: {"ok":true}]`;

exports[`A Client Should match the snapshot: sample client 1`] = `
Client {
  "axiosInstance": Object {
    "create": [Function],
    "nextResponse": Object {},
    "request": [Function],
  },
  "base": [Function],
  "debug": false,
  "globalResponses": Array [
    ApiResponse {
      "model": Model {
        "base": InterfaceType {
          "_tag": "InterfaceType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "{ ok: false }",
          "props": Object {
            "ok": LiteralType {
              "_tag": "LiteralType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "false",
              "validate": [Function],
              "value": false,
            },
          },
          "validate": [Function],
        },
        "name": "API 500 resource",
        "operation": "",
        "samples": Object {},
        "schema": InterfaceType {
          "_tag": "InterfaceType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "{ ok: false }",
          "props": Object {
            "ok": LiteralType {
              "_tag": "LiteralType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "false",
              "validate": [Function],
              "value": false,
            },
          },
          "validate": [Function],
        },
      },
      "status": 500,
    },
  ],
  "resources": Object {
    "extraResource": Resource {
      "basePath": "",
      "operations": Object {
        "getOk": Operation {
          "method": "GET",
          "name": "getOk",
          "responses": Array [
            ApiResponse {
              "model": Model {
                "base": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ ok: boolean }",
                  "props": Object {
                    "ok": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
                "name": "API sample resource",
                "operation": "getOk",
                "samples": Object {},
                "schema": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ ok: boolean }",
                  "props": Object {
                    "ok": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
              },
              "status": 200,
            },
          ],
          "url": [Function],
        },
      },
    },
    "samples": Resource {
      "basePath": "/samples",
      "operations": Object {
        "getOk": Operation {
          "method": "GET",
          "name": "getOk",
          "responses": Array [
            ApiResponse {
              "model": Model {
                "base": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ ok: boolean }",
                  "props": Object {
                    "ok": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
                "name": "API sample resource",
                "operation": "getOk",
                "samples": Object {},
                "schema": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ ok: boolean }",
                  "props": Object {
                    "ok": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
              },
              "status": 200,
            },
          ],
          "url": [Function],
        },
        "getSample": Operation {
          "method": "GET",
          "name": "getSample",
          "options": Object {},
          "responses": Array [
            ApiResponse {
              "model": Model {
                "base": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ ok: boolean }",
                  "props": Object {
                    "ok": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
                "name": "API sample resource",
                "operation": "getOk",
                "samples": Object {},
                "schema": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ ok: boolean }",
                  "props": Object {
                    "ok": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
              },
              "status": 200,
            },
            ApiResponse {
              "model": Model {
                "base": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ accepted: boolean }",
                  "props": Object {
                    "accepted": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
                "name": "API sample resource",
                "operation": "getSample",
                "samples": Object {},
                "schema": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ accepted: boolean }",
                  "props": Object {
                    "accepted": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
              },
              "status": 202,
            },
          ],
          "url": [Function],
        },
        "withMock": Operation {
          "method": "GET",
          "mock": [Function],
          "name": "withMock",
          "responses": Array [
            ApiResponse {
              "model": Model {
                "base": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ ok: boolean }",
                  "props": Object {
                    "ok": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
                "name": "API sample resource",
                "operation": "getOk",
                "samples": Object {},
                "schema": InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ ok: boolean }",
                  "props": Object {
                    "ok": BooleanType {
                      "_tag": "BooleanType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "boolean",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
              },
              "status": 200,
            },
          ],
          "url": [Function],
        },
      },
    },
  },
  "strictTypes": false,
  "throwErrors": false,
}
`;
